homeassistant/components/mazda/sensor.py

```diff
@@ -64,28 +52,19 @@ def native_value(self):
 class MazdaFuelDistanceSensor(MazdaEntity, SensorEntity):
     """Class for the fuel distance sensor."""
 
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        vehicle_name = self.get_vehicle_name()
-        return f"{vehicle_name} Fuel Distance Remaining"
+    _attr_icon = "mdi:gas-station"
 
-    @property
-    def unique_id(self):
-        """Return a unique identifier for this entity."""
-        return f"{self.vin}_fuel_distance_remaining"
+    def __init__(self, client, coordinator, index) -> None:
+        """Initialize Mazda fuel distance sensor."""
+        super().__init__(client, coordinator, index)
 
-    @property
-    def native_unit_of_measurement(self):
-        """Return the unit of measurement."""
-        if self.hass.config.units.name == CONF_UNIT_SYSTEM_IMPERIAL:
-            return LENGTH_MILES
-        return LENGTH_KILOMETERS
-
-    @property
-    def icon(self):
-        """Return the icon to use in the frontend."""
-        return "mdi:gas-station"
+        self._attr_name = f"{self.vehicle_name} Fuel Distance Remaining"
+        self._attr_unique_id = f"{self.vin}_fuel_distance_remaining"
+        self._attr_native_unit_of_measurement = (
+            LENGTH_KILOMETERS
+            if coordinator.hass.config.units.is_metric
+            else LENGTH_MILES
+        )
```

Comment:
This looks like an accurate translation of what was there before, but I think what was there before may not be correct. Native unit of measurement is supposed to be the unit that is returned from native_value; Home Assistant will convert that to whatever unit is appropriate for the user's configuration. So looking at the user's configuration here seems fishy. You should probably confine this PR to switching to attributes, but it would be worth investigating whether this is in fact causing issues and submitting another PR with a fix if so.

Reply:
I believe that HA only converts units when the `native_unit_of_measurement` is °C or °F and the `device_class` is temperature (this is documented [here](https://developers.home-assistant.io/docs/core/entity/sensor#properties)). I just did a test and changed `native_unit_of_measurement` to always return `LENGTH_KILOMETERS`, and `native_value` to always return the value in kilometers, but HA did not convert it to miles despite being configured to use imperial units.

Since we are checking the user's unit configuration both in `native_unit_of_measurement` and `native_value`, the number being displayed will always match the unit being displayed. So I don't think there is an issue here, but please let me know if I'm missing something.

Reply:
Well, @jkeljo is correct. The unit of measurement for this entity changes based on the unit system configured in Home Assistant, not the device that is integrated.

This is incorrect behavior. It should rely on its own data.

Reply:
@frenck Thanks for clarifying. What would be the best course of action here? I'm in the US and we use miles for car odometers. If I changed the odometer entity to always use kilometers regardless of the unit system preference, I think that be a negative user experience for a lot of people.

Reply:
Oh, right I see! It is converted based on this value.
Sorry I missed that part 👍